Buys a product (from a category) triggers [MAILPOET-5516] [MAILPOET-5517]

websupporter opened the pull request:
## Description
Introduces two new triggers:
* Triggers when a product has been purchased
* Triggers when a product from a category has been purchased

## Code review notes
* Utilizes the backend logic for the filters to evaluate whether the trigger should be run.
* The `createRootURLMiddleware` had to be extended to take another route when `wc-analytics` API queries are made.
* Cleans up some wrong text domains in JavaScript.
* The triggers will run only once per order (see `isTriggeredBy()`) and the editor can select a specific order status. Sometimes you might want to run the automation only on completed, sometimes as soon as possible.

## QA notes
Variable products do not properly work because of [MAILPOET-5553]

## Linked PRs

_N/A_

## Linked tickets
[MAILPOET-5516] [MAILPOET-5517]

## After-merge notes

_N/A_

## Tasks

- [x] I followed [best practices](https://codex.wordpress.org/I18n_for_WordPress_Developers) for translations
- [x] I added sufficient test coverage
- [x] I embraced TypeScript by either creating new files in TypeScript or converting existing JavaScript files when making changes


[MAILPOET-5553]: https://mailpoet.atlassian.net/browse/MAILPOET-5553?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAILPOET-5516]: https://mailpoet.atlassian.net/browse/MAILPOET-5516?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ
[MAILPOET-5517]: https://mailpoet.atlassian.net/browse/MAILPOET-5517?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

websupporter committed:
Fix text-domains

[MAILPOET-5516]

websupporter committed:
Switch api URL for wc-analytics

[MAILPOET-5516]

websupporter committed:
Prepare <OrderStatusPanel /> to be used by different steps

[MAILPOET-5516]

websupporter committed:
Add Buys A Product trigger frontend

[MAILPOET-5516]

websupporter committed:
Make method public

[MAILPOET-5516]

websupporter committed:
Add BuysAProduct trigger backend

[MAILPOET-5516]

websupporter committed:
Add test for BuysAProductTrigger

[MAILPOET-5516]

websupporter committed:
Add BuysFromACategory Trigger frontend

[MAILPOET-5517]

websupporter committed:
Add BuysFromACategory Trigger backend

[MAILPOET-5517]

websupporter committed:
Add test for BuysFromACategoryTrigger

[MAILPOET-5517]

rodrigoprimo commented on mailpoet/assets/js/src/automation/api/index.ts:
```diff
@@ -15,6 +15,12 @@ export type ApiError = {
 
 export const initializeApi = () => {
-  apiFetch.use(apiFetch.createRootURLMiddleware(apiUrl));
+  apiFetch.use((options, next) => {
```
Maybe it is because I'm not super familiar with the automations code, but it is not clear to me from the code changes, the commit message and the PR description why this change is needed and how it relates to the triggers that were added in this PR. Could you please elaborate on that? For future commits, I would suggest documenting the why of the change in the commit message.

websupporter replied:
We use the search component of WooCommerce now for searching products and categories:
https://github.com/mailpoet/mailpoet/pull/5162/commits/792ef36e7bd893318a0fca4676cc83e6b334db06#diff-21f48913c311bceda3481eef0550b8886d99c8383c75864d48e8fd59746b27abR2

`<Search />` will do its own requests through `apiFetch`. Without this change the URL would be `wp-json/mailpoet/wc-analytics/...` instead of `wp-json/wc-analytics`.

rodrigoprimo replied:
Thanks for the additional details.

rodrigoprimo commented on mailpoet/assets/js/src/automation/integrations/woocommerce/index.tsx:
```diff
@@ -2,11 +2,13 @@ import { registerStepType } from '../../editor/store';
 import { step as OrderStatusChanged } from './steps/order_status_changed';
 import { step as AbandonedCartTrigger } from './steps/abandoned_cart';
 import { MailPoet } from '../../../mailpoet';
+// Insert new imports here
```
This comment and the comment below is needed because the order of the imports and the order in which the steps are registered matter? If that is the case, maybe we should make it explicit in the comments?

websupporter replied:
No. tbh, its a residue of https://github.com/mailpoet/mailpoet/pull/5144/commits/51dfbbb1a723232a206da39710201c11f2351ec5 from another PR. This commit adds a roboli command to quickly create a new trigger. And these are basically placeholders where the command then inserts the new trigger. I used the commit in the branch to have my command and later removed the commit, but these comments remained for some reason. I would just leave it for now as it is.

rodrigoprimo replied:
Ok

rodrigoprimo reviewed (commented):
Overall looks good to me.

I left a few inline comments and besides those, I have a few questions and one potential issue:

- I haven't checked, but I believe that external/affiliate products don't trigger the `woocommerce_order_status_changed` action, so maybe they should not be included in the list of products that can be used in the trigger?
- I wonder if we should indicate somehow that we are waiting for an ajax response when searching for products?
- I created an automation using the `buys a product` trigger and the `send email` action. It worked fine a few times, but I got the following error on some occasions. It seems to me that the error will happen for every email that is not part of the WooCommerce Customers list. Are you able to reproduce this error? I'm not sure if this is something directly related to this PR or not. The error is logged to wp_mailpoet_automation_run_logs.

```
{"message":"Subscriber ID '189' is not subscribed to segment ID '2'.","errorClass":"MailPoet\\InvalidStateException","code":0,"trace":[{"file":"/var/www/html/wp-content/plugins/mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendEmailAction.php","line":146,"function":"create","class":"MailPoet\\Exception","type":"::","args":[]},{"file":"/var/www/html/wp-content/plugins/mailpoet/lib/Automation/Engine/Control/Steps/ActionStepRunner.php","line":28,"function":"run","class":"MailPoet\\Automation\\Integrations\\MailPoet\\Actions\\SendEmailAction","type":"->","args":[{}]},{"file":"/var/www/html/wp-content/plugins/mailpoet/lib/Automation/Engine/Control/StepHandler.php","line":166,"function":"run","class":"MailPoet\\Automation\\Engine\\Control\\Steps\\ActionStepRunner","type":"->","args":[{},{}]},{"file":"/var/www/html/wp-content/plugins/mailpoet/lib/Automation/Engine/Control/StepHandler.php","line":111,"function":"handleStep","class":"MailPoet\\Automation\\Engine\\Control\\StepHandler","type":"->","args":[{}]},{"file":"/var/www/html/wp-includes/class-wp-hook.php","line":310,"function":"handle","class":"MailPoet\\Automation\\Engine\\Control\\StepHandler","type":"->","args":[{"automation_run_id":4,"step_id":"1ae2174dea21c2e2"}]},{"file":"/var/www/html/wp-includes/class-wp-hook.php","line":334,"function":"apply_filters","class":"WP_Hook","type":"->","args":["",[{"automation_run_id":4,"step_id":"1ae2174dea21c2e2"}]]},{"file":"/var/www/html/wp-includes/plugin.php","line":565,"function":"do_action","class":"WP_Hook","type":"->","args":[[{"automation_run_id":4,"step_id":"1ae2174dea21c2e2"}]]},{"file":"/var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/actions/ActionScheduler_Action.php","line":56,"function":"do_action_ref_array","args":["mailpoet/automation/step",[{"automation_run_id":4,"step_id":"1ae2174dea21c2e2"}]]},{"file":"/var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/abstracts/ActionScheduler_Abstract_QueueRunner.php","line":88,"function":"execute","class":"ActionScheduler_Action","type":"->","args":[]},{"file":"/var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_QueueRunner.php","line":166,"function":"process_action","class":"ActionScheduler_Abstract_QueueRunner","type":"->","args":[8217,"WP Cron"]},{"file":"/var/www/html/wp-content/plugins/woocommerce/packages/action-scheduler/classes/ActionScheduler_QueueRunner.php","line":136,"function":"do_batch","class":"ActionScheduler_QueueRunner","type":"->","args":[25,"WP Cron"]},{"file":"/var/www/html/wp-includes/class-wp-hook.php","line":310,"function":"run","class":"ActionScheduler_QueueRunner","type":"->","args":["WP Cron"]},{"file":"/var/www/html/wp-includes/class-wp-hook.php","line":334,"function":"apply_filters","class":"WP_Hook","type":"->","args":["",["WP Cron"]]},{"file":"/var/www/html/wp-includes/plugin.php","line":565,"function":"do_action","class":"WP_Hook","type":"->","args":[["WP Cron"]]},{"file":"/var/www/html/wp-cron.php","line":191,"function":"do_action_ref_array","args":["action_scheduler_run_queue",["WP Cron"]]}]}
```

Here is the line that generates this error: https://github.com/mailpoet/mailpoet/blob/0cc02b38a59cfacc5ad8a78999b3f4ce5ebdf2dd/mailpoet/lib/Automation/Integrations/MailPoet/Actions/SendEmailAction.php#L146

websupporter commented:
> I wonder if we should indicate somehow that we are waiting for an ajax response when searching for products?

Yeah, I would like that too, but I did not find a way to do that in the `<Search />` component :/

>  believe that external/affiliate products don't trigger the woocommerce_order_status_changed action, so maybe they should not be included in the list of products that can be used in the trigger?

That would be odd. How do they work? Isn't an affilate product in the end just a link to the affiliate shop or something. What I mean, I wonder how they would end up in the cart and in an order... ahhh... and thus they should not be part of the `<Search />` result :thinkingface:

I am on support rotation this week, so I open a ticket to suggest an improvement [MAILPOET-5594]

> I got the following error on some occasions

Atm, I think that is expected (we cant send emails to nonsubscribed subscribers), but in any case, it would be unrelated to this PR (you would produce the same result with e.g. the order status change trigger) but to the behavior defined in the send email step.

[MAILPOET-5594]: https://mailpoet.atlassian.net/browse/MAILPOET-5594?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

rodrigoprimo commented:
> Yeah, I would like that too, but I did not find a way to do that in the <Search /> component :/

:+1: 

> I am on support rotation this week, so I open a ticket to suggest an improvement [MAILPOET-5594](https://mailpoet.atlassian.net/browse/MAILPOET-5594?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ)

:+1: 

> Atm, I think that is expected (we cant send emails to nonsubscribed subscribers), but in any case, it would be unrelated to this PR (you would produce the same result with e.g. the order status change trigger) but to the behavior defined in the send email step.

Ok, thanks for confirming that this behavior is not directly related to this PR. Maybe we should change the error message to make it more clear to the users that this is an expected failure? I wonder if we should go even further and change the status to something other than failure. What about creating a separate Jira ticket to look into this?

[MAILPOET-5594]: https://mailpoet.atlassian.net/browse/MAILPOET-5594?atlOrigin=eyJpIjoiNWRkNTljNzYxNjVmNDY3MDlhMDU5Y2ZhYzA5YTRkZjUiLCJwIjoiZ2l0aHViLWNvbS1KU1cifQ

websupporter committed:
Change method name to withStep

trunk changed the method name, this commit reflects this change for the tests

[MAILPOET-5516]

websupporter committed:
Ensure options.path exists

[MAILPOET-5516]

Aschepikov reviewed (approved):
GJ!
